### PR TITLE
Add IsValidSnippetInsertionAtPosition support

### PR DESCRIPTION
**Pull request by jramsay**
Add IsValidSnippetInsertionAtPosition support so that hosts can determine if snippet entries should be inserted in the completion list

### CODE REVIEW

**Comment by jramsay** [comment]
// cc: @mhegazy, @vladima , @billti , @zhengbli

**Comment by zhengbli** [comment]
It looks like the editor needs to make two calls whenever the it needs to provide completion list: one for the snippet insertion and the other for the actual completion. Why not just combine these two and make the `shouldInsertSnippet` a property for the completion list response?

**Comment by vladima** [comment]
it is not useful for VS it its current state  when there are two separate completion providers for regular completions and for snippets. In theory we can combine them into one however before doing I'd like to see how much time it takes to execute `isValidSnippetInsertion`

File: src/services/services.ts [hunk 1]
```diff
@@ -1665,6 +1665,11 @@ namespace ts {
             return true;
         }
 
+        function isValidSnippetInsertionAtPosition(fileName: string, position: number): boolean {
+            const completionInfo = getCompletionsAtPosition(fileName, position);
```
Comment (vladima) [inline src/services/services.ts]: current implementation is effectively computes set of completions for the second time which is a bit too heavyweight given that it should be possible to provide answer for the question `isValidSnippetInsertionAtPosition` using only syntactic information and don't touch semantic data at all.

Comment (jramsay) [inline src/services/services.ts]: After discussion with @vladima & @mhegazy it makes sense to merge our regular & snippet completion providers in VS as it is not worth the effort to extract the syntactic information to determine this on the script side.  Closing this review out and will follow up with a separate review on the VS host side.

Comment (jramsay) [inline src/services/services.ts]: After discussion with @vladima & @mhegazy it makes sense to merge our regular & snippet completion providers in VS as it is not worth the effort to extract the syntactic information to determine this on the script side.  Closing this review out and will follow up with a separate review on the VS host side.

**Comment by jramsay** [comment]
After discussion with @vladima & @mhegazy it makes sense to merge our regular & snippet completion providers in VS as it is not worth the effort to extract the syntactic information to determine this on the script side. Closing this review out and will follow up with a separate review on the VS host side.

**Pull request closed without merging**